doc/design.md

```diff
@@ -0,0 +1,258 @@
+# Design
+
+
+## Introduction
+This document describes the design of the advanced analytics framework on which 
+complex data analysis algorithms can be built.
+
+## Constraints / Limitations
+
+- The best way to run dynamic SQL queries is to use Lua scripts. Because Lua
+scripting enables us to run queries in the same transaction using the `pquery` method.
+- The results of `pquery` are simple handles and there is a limit of 250 open 
+result set handles for Lua scripts.
+- The main logic of the framework should be implemented in Python, because it is 
+widely used in data processing and analysis.
+- Python UDF, where the main logic is placed, returns the result of the query 
+and finishes its execution. Due to that, it cannot be kept running and used 
+as server.
+
+
+## Assumptions
+
+- Since the number open result set handles in Lua scripts is limited, it is 
+assumed that Lua scripts is not nested more than 250 times.
+- The result of `pquery` is assumed to not exceed the db memory limit in size. 
+The DB usually allows 4 GB per query and node. The UDF can consume up to 32 GB 
+per node, but gets throttled down after the 4 GB. However, the 32 GB are shared 
+between all queries on a node.
+- The state of the algorithm is not too large.
```

Comment:
```suggestion
- Because Python UDF can't run constantly, we need to save its state to the BucketFS. Saving and loading needs to reasonable fast to not introduce too much overhead. For that, we assume the state of the Python UDF is reasonable small (max a few MB).
```